### DISCUSSION TITLE
A way of triggering a pipeline

### DISCUSSION THREAD

**Discussion by northern-mindset** [vitaly-t/iter-ops#230] [Q&A]
Suppose I'm building a pipeline which does not provide any expected result, i.e. it is an iterator of `void`.

What is the supposed/idiomatic way of triggering the pipeline in a functional way? I mean, not by `for … of`.

From the documentation, I see only `drain()` plus `.first`.

**Comment by vitaly-t** [comment]
If it is an iterator of `void`, it has no data, so what is triggering an unforeknown sequence is supposed to yield in your view?

P.S. Accessing `.first` always triggers the pipeline, but if it has no data, you will get no data.

**Reply by northern-mindset** [reply]
Perhaps I was not clear about “no expected result”. I mean that the pipeline yields `void` values because all the processing is done inside the pipeline. Basically, I'm (ab)using iterators for processing data, and there is no output data to collect. But still I need to exhaust the whole pipeline.

Simplified pseudo-code:
```js
pipeAsync(
	someSourceIterator,
      map(...), // Here some async processes are launched, map yields `Promise<void>`.
      wait()
    )
```

**Reply by vitaly-t** [reply]
Are you trying to get something yielded when there is no data? For that, there is [defaultEmpty](https://vitaly-t.github.io/iter-ops/functions/defaultEmpty.html) operator.

**Reply by northern-mindset** [reply]
No, there is data initially. Then this data is processed asynchronously, values are transformed into `Promise<void>` and awaited using `wait()` operation. Thus, the pipeline yields a series of `void`s.

See the first example in https://vitaly-t.github.io/iter-ops/functions/wait.html. My case is almost same, just promises are of `void`. Therefore I don't want/need a `for ... of` loop to run the pipeline. I just want to run it.

Am I doing something stupid or misunderstanding anything?

**Reply by vitaly-t** [reply]
It's alright, just do await `.first` to trigger the pipeline, and it should be fine ;)

**Reply by northern-mindset** [reply]
All right. With `drain()`, right?

**Reply by vitaly-t** [reply]
Whatever works for you, I guess :) Also, check out custom [mapWait](https://github.com/vitaly-t/iter-ops-extras/blob/main/src/map-wait.ts) ;)
